### PR TITLE
Ensure we either report or crash fatal errors

### PR DESCRIPTION
**Pull request by imbstack**
This should ensure that we report fatal errors and crash. It may fail to report the error but at least it will crash!

Upon writing this, I've realized the reason that dependencyResolver isn't crashing! It's because the 'error' we're reporting is turned into an unreseolved promise rather than an uncaught error and we don't crash on unresolved promises! We should throw the error inside a `process.nextTick()` or  find some other way to deal with this if we want to crash on these errors.

### CODE REVIEW

**Comment by imbstack** [comment]
Well, let me make these tests pass first, I guess.

**Comment by jonasfj** [comment]
> Upon writing this, I've realized the reason that dependencyResolver isn't crashing! It's because the 'error' we're reporting is turned into an unreseolved promise rather than an uncaught error and we don't crash on unresolved promises!

I don't follow this logic... We do `.catch(err => this.emit('error', err))`

File: src/monitor.js [hunk 1]
```diff
@@ -33,14 +33,23 @@ class Monitor {
...
       sentry.client.captureException(err, {
         tags: _.defaults({
           prefix: this._opts.project + (this._opts.prefix || '.root'),
           process: this._opts.process || 'unknown',
         }, tags),
         level,
       });
+
+      return new Promise((accept, reject) => {
+        sentry.client.once('logged', () => {
```
Comment (jonasfj) [inline src/monitor.js]: I can't imagine this will work well with multiple parallel errors being sent...

**Comment by imbstack** [comment]
Does http://stackoverflow.com/questions/30715367/why-can-i-not-throw-inside-a-promise-catch-handler make sense to you? It explained what was happening to me fairly well I think.

I built a test case to confirm that it's how this works and tried it out.

File: src/monitor.js (hunk 1 above)
Comment (jonasfj) [inline src/monitor.js]: But given that we only care about whether it was sent or not upon crashing... Maybe we only look at these events in the special crash-handler...

Comment (jonasfj) [inline src/monitor.js]: ps. wow this has so many corner cases...

Comment (imbstack) [inline src/monitor.js]: Haha, yes. That's why I added the timeout to try to ensure that at the very least we crash.

**Comment by jonasfj** [comment]
> I built a test case to confirm that it's how this works and tried it out.

Fine I'll build a counter case :)
`this.emit('error', err)` does not throw an exception afaik.

File: src/monitor.js [hunk 2]
```diff
@@ -33,14 +33,23 @@ class Monitor {
...
+          accept('Succesfully logged error to Sentry.');
```
Comment (jonasfj) [inline src/monitor.js]: No reason to send an argument... Usually:

``` js
return new Promise((accept, reject) => {
  sentry.client.once('logged', accept);
  sentry.client.once('error', reject);
});
```

But even this leaks... Sorry...

We probably have to do:

``` js
return new Promise((accept, reject) => {
  let onLogged, onError;
  onLogged = () => {
    sentry.client.removeListener('error', onError);
    accept();
  };
  onError = (err) => {
    sentry.client.removeListener('logged', onLogged);
    reject(err);
  }
  sentry.client.once('logged', onLogged);
  sentry.client.once('error', onError);
});
```

Yes, this sucks... But it doesn't leak memory... hopefully this doesn't matter... but if one day we have 100k internal errors, it would suck if we also got a memory consumption issue :)

File: src/monitor.js [hunk 3]
```diff
@@ -33,14 +33,23 @@ class Monitor {
...
+        });
+        sentry.client.once('error', (e) => {
+          reject('Failed to log error to Sentry: ', e.reason);
```
Comment (jonasfj) [inline src/monitor.js]: Actually, creating a custom error message like this is nice...

But please send error objects back when doing this... Also are you sure there is always an `e.reason` property...
Also `reject` only takes a single parameter afaik...

I would just do: `reject(new Error("Failed sending to sentry: " + err))` or maybe err.toString() which might be the same...

**Comment by jonasfj** [comment]
@imbstack you are right about .emit('error', I read the node docs wrong..

IMO, let's fix the memory leaks and then land this...

File: src/monitor.js (hunk 3 above)
Comment (imbstack) [inline src/monitor.js]: It should always have the `e.reason`. https://docs.getsentry.com/hosted/clients/node/usage/#events

I'll make the other changes you mention shortly.

File: src/monitor.js (hunk 2 above)
Comment (imbstack) [inline src/monitor.js]: I thought using `.once` was designed to remove the listener after it fired. I'm probably misunderstanding this and will read the docs for a bit.

Comment (imbstack) [inline src/monitor.js]: Oh wait, I see why this is needed. Wow that's gross. I'll add it shortly.

**Comment by imbstack** [comment]
@jonasfj ok, this should be ready to look over again.

File: src/monitor.js (hunk 2 above)
Comment (jonasfj) [inline src/monitor.js]: yeah, it not very nice :)

File: src/monitor.js [hunk 4]
```diff
@@ -212,10 +235,22 @@ async function monitor(options) {
   }
 
   if (opts.patchGlobal) {
-    process.on('uncaughtException', (err) => {
+    process.on('uncaughtException', async (err) => {
+      console.log('Uncaught Exception! Attempting to report to Sentry and crash.');
       console.log(err.stack);
```
Comment (jonasfj) [inline src/monitor.js:240]: I would also like: `console.log("error properties: %j", err)` To get a dictionary of properties on the error object.
